plugins/kotlin-as-java/src/main/kotlin/converters/KotlinToJavaConverter.kt

```diff
@@ -259,24 +262,51 @@ internal fun DClass.asJava(): DClass = copy(
             )
         }, // name may not always be valid here, however classNames should always be not null
     functions = functionsInJava(),
-    properties = properties
-        .filterNot { it.hasJvmSynthetic() }
-        .map { it.asJava() },
-    classlikes = classlikes.map { it.asJava() },
+    properties = propertiesInJava(),
+    classlikes = classlikesInJava(),
     generics = generics.map { it.asJava() },
+    companion = companion.companionAsJava(),
     supertypes = supertypes.mapValues { it.value.map { it.asJava() } },
     modifier = if (modifier.all { (_, v) -> v is KotlinModifier.Empty }) sourceSets.associateWith { JavaModifier.Final }
     else sourceSets.associateWith { modifier.values.first() }
 )
 
+/**
+ * Companion objects requires some custom logic for rendering as Java.
+ * They are excluded from usual classlikes rendering and added after.
+ */
+internal fun DClass.classlikesInJava(): List<DClasslike> {
+    val classlikes = classlikes
+        .filter { it.name != companion?.name }
+        .map { it.asJava() }
+
+    val companionAsJava = companion.companionAsJava()
+    return if (companionAsJava != null) classlikes.plus(companionAsJava) else classlikes
+}
+
+
 internal fun DClass.functionsInJava(): List<DFunction> =
     properties
-        .filter { it.jvmField() == null && !it.hasJvmSynthetic() }
+        .filter { !it.isJvmField && !it.hasJvmSynthetic() }
         .flatMap { property -> listOfNotNull(property.getter, property.setter) }
         .plus(functions)
+        .plus(companion.staticFunctionsForJava())
         .filterNot { it.hasJvmSynthetic() }
         .flatMap { it.asJava(it.dri.classNames ?: it.name) }
 
+internal fun DClass.propertiesInJava(): List<DProperty> {
+    val propertiesFromCompanion = companion
+        .staticPropertiesForJava()
+        .filterNot { it.hasJvmSynthetic() }
+        .map { it.asJava(isFromObjectOrCompanion = true) }
+    val companionInstanceProperty = listOfNotNull(companion.companionInstancePropertyForJava())
+    val ownProperties = properties
+        .filterNot { it.hasJvmSynthetic() }
+        .map { it.asJava() }
+
+    return propertiesFromCompanion + ownProperties + companionInstanceProperty
```

Comment:
```suggestion
    val companionInstanceProperty = companion.companionInstancePropertyForJava()
    val ownProperties = properties
        .filterNot { it.hasJvmSynthetic() }
        .map { it.asJava() }

    return propertiesFromCompanion + ownProperties + (companionInstanceProperty ?: emptyList())
```

Reply:
Not sure that the suggested change is valid. Since `companionInstanceProperty ` has the type `DProperty`  and the `emptyList` is a `List` of something then the expression in braces has the type `Any`.

Reply:
Sorry, I should've given more context for why my proposed change

The main reason is the name of the val. When I see `companionInstanceProperty`, judging solely by the name, I expect it to be a single object, not a list, especially not an empty one. This doesn't hurt much in this context, but if someone were to add more code that would involve using `companionInstanceProperty`, it could lead to some ugly solutions

There's also a very minor optimization that comes as a nice bonus (avoid creating an extra list and an array for vararg), but it's not substantial and hopefully escape analysis can take care of it

But it's a minor thing at this point, I don't mind it being the way it is now, so it's up to you

Reply:
Oh, I see what you meant, `(companionInstanceProperty ?: emptyList())` does not compile, sorry. The suggestion can be dissmissed then. The main point was about the name, so it can probably be inlined to `propertiesFromCompanion + ownProperties + listOfNotNull(companionInstanceProperty)`